src/pac.c

```diff
@@ -124,13 +144,47 @@ pac_alloc_real(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, size_t size,
 		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_muzzy,
 		    NULL, size, alignment, zero, guarded);
 	}
+#ifdef LIMIT_USIZE_GAP
+	/*
+	 * Split is not supported when (!map_coalesce && !opt_retain).
+	 * Do not try to bump the size for this case.
+	 */
+	if (sz_limit_usize_gap_enabled() && edata == NULL &&
+	    (maps_coalesce || opt_retain)) {
+		size_t caching_size = pac_batched_alloc_retained_size(
+		    size);
+		edata = ecache_alloc_grow(tsdn, pac, ehooks,
+		    &pac->ecache_retained, NULL, caching_size,
+		    alignment, zero, guarded);
+
+		if (edata != NULL && caching_size > size) {
+			edata_t *trail = extent_split_wrapper(tsdn, pac,
+			    ehooks, edata, size, caching_size - size,
+			    /* holding_core_locks */ false);
+			if (trail == NULL) {
+				ecache_dalloc(tsdn, pac, ehooks,
+				    &pac->ecache_retained, edata);
+				edata = NULL;
+			} else {
+				ecache_dalloc(tsdn, pac, ehooks,
+				    &pac->ecache_dirty, trail);
+			}
+		}
+
+		if (config_stats && edata != NULL) {
```

Comment:
This is redundant code. We need to do better and avoid this.

bool newly_mapped = false;
...

Then here and the next block mark it true, and check it to update the counter by the end.

Reply:
Modified.